spec/services/notification_service/course_subjects_updated_spec.rb

```diff
@@ -0,0 +1,155 @@
+      let(:previous_subject_names) { ["primary with english"] }
+      let(:updated_subject_names) { ["primary with mathematics"] }
+      let(:previous_course_name) { previous_subject_names.first }
+      let(:updated_course_name) { updated_subject_names.first }
+
+      let(:subscribed_user) { create(:user) }
+      let(:non_subscribed_user) { create(:user) }
+      let(:user_subscribed_to_other_provider) { create(:user) }
+
+      let(:subscribed_notification) do
+        create(
+          :user_notification,
+          user: subscribed_user,
+          course_update: true,
+          provider_code: accredited_body.provider_code,
+        )
+      end
+
+      let(:non_subscribed_notification) do
+        create(
+          :user_notification,
+          user: non_subscribed_user,
+          course_update: false,
+          provider_code: accredited_body.provider_code,
+        )
+      end
+
+      let(:other_provider_notification) do
+        create(
+          :user_notification,
+          user: user_subscribed_to_other_provider,
+          course_publish: true,
+          provider_code: other_accredited_body.provider_code,
+        )
+      end
+      let(:self_accredited) { false }
+      let(:findable) { true }
+
+      def setup_notifications
+        allow(CourseSubjectsUpdatedEmailMailer).to receive(:course_subjects_updated_email).and_return(double(deliver_later: true))
+        subscribed_notification
+        non_subscribed_notification
+        other_provider_notification
+        allow(course).to receive(:self_accredited?).and_return(self_accredited)
+        allow(course).to receive(:findable?).and_return(findable)
+      end
+
+      context "with a course that is in the current cycle" do
+        before { setup_notifications }
+
+        it "sends notifications" do
+          expect(CourseSubjectsUpdatedEmailMailer).to receive(:course_subjects_updated_email)
+          expect(course.recruitment_cycle).to eql(RecruitmentCycle.current)
+          described_class.call(
+            course: course,
+            previous_subject_names: previous_subject_names,
+            previous_course_name: previous_course_name,
+          )
+        end
+      end
+
+      context "with a course that is not in the current cycle" do
+        let(:provider) { create(:provider, :next_recruitment_cycle) }
+        let(:course) { create(:course, accredited_body_code: accredited_body.provider_code, provider: provider) }
+
+        before { setup_notifications }
+
+        it "does not send a notification" do
+          expect(CourseSubjectsUpdatedEmailMailer).to_not receive(:course_subjects_updated_email)
+          expect(course.recruitment_cycle).to_not eql(RecruitmentCycle.current)
+          described_class.call(
+            course: course,
+            previous_subject_names: previous_subject_names,
+            previous_course_name: previous_course_name,
+          )
+        end
+      end
```

Comment:
I'm finding these tests really hard to read. Now that I've been staring at them for a bit I feel like I can kind of understand what's being tested, but I can't help but wonder if it wouldn't be better organised if we used the pattern where we do call the service in a before block, and then do the setup and testing separately. Something like:

```ruby
    describe "#call" do
      # all the let blocks

      before do
        setup_notifications

        described_class.call(
          course: course,
          previous_subject_names: subject_names,
          previous_course_name: course_name,
        )
      end

      describe CourseSubjectsUpdatedEmailMailer do
        before do
          allow(CourseSubjectsUpdatedEmailMailer).to receive(:course_subjects_updated_email)
        end

        context "with a course that is in the current cycle" do
          it { is_expected.to receive(:course_subjects_updated_email) }
        end

        context "with a course that is not in the current cycle" do
          let(:provider) { create(:provider, :next_recruitment_cycle) }
          let(:course) { create(:course, accredited_body_code: accredited_body.provider_code, provider: provider) } 

          it { is_expected.to_not receive(:course_subjects_updated_email) }
        end
      end

      # and the rest of the specs modified similarly
```

Reply:
Alternatively, it seems that this code is used each time to trigger the work needed for the test:

```ruby
described_class.call(
  course: course,
  previous_subject_names: previous_subject_names,
  previous_course_name: previous_course_name,
)
```

I find it's hard to parse that while also keeping all the context in my head ... so maybe turn that into a method that makes is clear that it's the same call throughout:

```ruby
def call_service
  described_class.call(
    course: course,
    previous_subject_names: previous_subject_names,
    previous_course_name: previous_course_name,
  )
end


it "does not send a notification" do
  expect(CourseSubjectsUpdatedEmailMailer).to_not receive(:course_subjects_updated_email)
  expect(course.recruitment_cycle).to_not eql(RecruitmentCycle.current)
  call_service
end 
```

Reply:
Yep cool - have made the changes